File: scripts/version-bump-check.py
#!/usr/bin/env python3
"""scripts/version-bump-check.py — flag plugins changed without a version bump.

Guards against the "silent no-update" trap. Claude Code resolves a plugin's
version from the first of: plugin.json `version`, the marketplace entry
`version`, then the git commit SHA. When `version` is set in plugin.json,
pushing new commits *without changing that string does nothing for existing
users* — Claude Code sees the same version and keeps the cached copy. So if a
plugin's shipped files change but `plugins/<name>/.claude-plugin/plugin.json`
`version` stays the same, users never receive the update.

This is the local mirror of the PR gate in
`.github/workflows/version-bump.yml`, so contributors can catch the problem
before pushing instead of waiting for CI.

Usage:
  python3 scripts/version-bump-check.py [BASE [HEAD]]

    BASE   git ref for the "before" side. Default: origin/main if it
           resolves, else HEAD.
    HEAD   git ref for the "after" side. Default: the working tree
           (so uncommitted edits count).

Exit codes:
  0  ok — every touched plugin bumped its version (or is new / unversioned)
  1  at least one plugin changed without a version bump
  2  setup error (git diff failed, etc.)

Plugins with no `version` field at all are skipped: they are intentionally
SHA-versioned, so every commit already reaches users on the next refresh.
Brand-new plugins (no version at BASE) are skipped too.
"""
from __future__ import annotations

import json
import re
import subprocess
import sys

PLUGIN_DIR_RE = re.compile(r"^(plugins/[^/]+)/")


def git(*args: str) -> subprocess.CompletedProcess[str]:
    return subprocess.run(["git", *args], capture_output=True, text=True)


def resolve_base(explicit: str | None) -> str:
    if explicit:
        return explicit
    if git("rev-parse", "--verify", "--quiet", "origin/main").returncode == 0:
        return "origin/main"
    return "HEAD"


def version_at(ref: str | None, path: str) -> str | None:
    """`version` of plugin.json at a ref, or None if the file/field is absent.

    ref=None reads the working-tree file.
    """
    if ref is None:
        try:
            with open(path, encoding="utf-8") as f:
                data = json.load(f)
        except (OSError, json.JSONDecodeError):
            return None
    else:
        r = git("show", f"{ref}:{path}")
        if r.returncode != 0:
            return None
        try:
            data = json.loads(r.stdout)
        except json.JSONDecodeError:
            return None
    v = data.get("version")
    return str(v) if v is not None else None


def main() -> int:
    base = resolve_base(sys.argv[1] if len(sys.argv) > 1 else None)
    head = sys.argv[2] if len(sys.argv) > 2 else None  # None => working tree

    if git("rev-parse", "--is-inside-work-tree").returncode != 0:
        print("version-bump-check: not a git repository; skipping")
        return 0

    diff_args = ["diff", "--name-only", base]
    if head:
        diff_args.append(head)
    diff_args += ["--", "plugins"]
    r = git(*diff_args)
    if r.returncode != 0:
        print(f"version-bump-check: git diff failed ({r.stderr.strip()})", file=sys.stderr)
        return 2

    changed = [ln for ln in r.stdout.splitlines() if ln.strip()]
    if not head:  # include untracked new files when comparing the working tree
        u = git("ls-files", "--others", "--exclude-standard", "--", "plugins")
        changed += [ln for ln in u.stdout.splitlines() if ln.strip()]

    touched = sorted(
        {m.group(1) for ln in changed if (m := PLUGIN_DIR_RE.match(ln))}
    )
    if not touched:
        print(f"version-bump-check: no plugin files changed vs {base}; nothing to enforce")
        return 0

    fail = 0
    for plugin in touched:
        manifest = f"{plugin}/.claude-plugin/plugin.json"
        before = version_at(base, manifest)
        after = version_at(head, manifest)
        if before is None:
            print(f"  new plugin (no base version): {plugin} — skipping")
            continue
        if after is None:
            print(f"  unversioned plugin (SHA-based updates reach users): {plugin} — skipping")
            continue
        if before == after:
            print(
                f"FAIL ({manifest}): version not bumped (still {before}); "
                f"bump it when changing files inside {plugin}/"
            )
            fail = 1
        else:
            print(f"  OK: {plugin} {before} -> {after}")

    if not fail:
        print("version-bump-check: all touched plugins bumped their version")
    return fail


if __name__ == "__main__":
    sys.exit(main())

File: scripts/ts/version-bump-check.ts
#!/usr/bin/env tsx
// Flags plugins changed without a version bump.
// Port of scripts/version-bump-check.py; output is byte-identical.
//
// Guards the "silent no-update" trap: Claude Code resolves a plugin's version
// from plugin.json `version` first, so shipping new files without changing that
// string leaves every existing user on the cached copy.
//
// Usage: tsx scripts/ts/version-bump-check.ts [BASE [HEAD]]
//   BASE defaults to origin/main when it resolves, else HEAD.
//   HEAD defaults to the working tree, so uncommitted edits count.
//
// Exit 0 ok · 1 a plugin changed without a bump · 2 setup error.
//
// Plugins with no `version` field are skipped - they are intentionally
// SHA-versioned, so every commit already reaches users. New plugins too.

import { spawnSync } from 'node:child_process';
import { readFileSync } from 'node:fs';

const PLUGIN_DIR_RE = /^(plugins\/[^/]+)\//;

const git = (...args: string[]) => spawnSync('git', args, { encoding: 'utf8' });

function resolveBase(explicit: string | undefined): string {
  if (explicit) return explicit;
  return git('rev-parse', '--verify', '--quiet', 'origin/main').status === 0 ? 'origin/main' : 'HEAD';
}

/** `version` of plugin.json at a ref, or null when file or field is absent.
 *  ref undefined reads the working tree. */
function versionAt(ref: string | undefined, path: string): string | null {
  let raw: string;
  if (ref === undefined) {
    try {
      raw = readFileSync(path, 'utf8');
    } catch {
      return null;
    }
  } else {
    const r = git('show', `${ref}:${path}`);
    if (r.status !== 0) return null;
    raw = r.stdout;
  }
  try {
    const v = (JSON.parse(raw) as { version?: unknown }).version;
    return v === undefined || v === null ? null : String(v);
  } catch {
    return null;
  }
}

const argv = process.argv.slice(2);
const base = resolveBase(argv[0]);
const head = argv[1]; // undefined => working tree

if (git('rev-parse', '--is-inside-work-tree').status !== 0) {
  console.log('version-bump-check: not a git repository; skipping');
  process.exit(0);
}

const diffArgs = ['diff', '--name-only', base];
if (head) diffArgs.push(head);
diffArgs.push('--', 'plugins');
const r = git(...diffArgs);
if (r.status !== 0) {
  console.error(`version-bump-check: git diff failed (${(r.stderr ?? '').trim()})`);
  process.exit(2);
}

const changed = (r.stdout ?? '').split('\n').filter((l) => l.trim());
if (!head) {
  // include untracked new files when comparing against the working tree
  const u = git('ls-files', '--others', '--exclude-standard', '--', 'plugins');
  changed.push(...(u.stdout ?? '').split('\n').filter((l) => l.trim()));
}

const touched = [
  ...new Set(changed.map((l) => PLUGIN_DIR_RE.exec(l)?.[1]).filter((x): x is string => Boolean(x))),
].sort();

if (touched.length === 0) {
  console.log(`version-bump-check: no plugin files changed vs ${base}; nothing to enforce`);
  process.exit(0);
}

let fail = 0;
for (const plugin of touched) {
  const manifest = `${plugin}/.claude-plugin/plugin.json`;
  const before = versionAt(base, manifest);
  const after = versionAt(head, manifest);
  if (before === null) {
    console.log(`  new plugin (no base version): ${plugin} — skipping`);
    continue;
  }
  if (after === null) {
    console.log(`  unversioned plugin (SHA-based updates reach users): ${plugin} — skipping`);
    continue;
  }
  if (before === after) {
    console.log(
      `FAIL (${manifest}): version not bumped (still ${before}); ` +
        `bump it when changing files inside ${plugin}/`,
    );
    fail = 1;
  } else {
    console.log(`  OK: ${plugin} ${before} -> ${after}`);
  }
}

if (!fail) console.log('version-bump-check: all touched plugins bumped their version');
process.exit(fail);
